File: LeetCodeRecord/MethodFirst/offer18-DeleteNode.java
package MethodFirst;

class DeleteNode {
    class ListNode {
        int val;
        ListNode next;

        ListNode(int x) {
            val = x;
            next = null;
        }
    }
    
    public ListNode deleteNode(ListNode head, int val) {
        ListNode prev = new ListNode(-1);
        ListNode bak = prev;
        prev.next = head;

        while (prev.next != null && prev.next.val != val) {
            prev = prev.next;            
        }

        if (prev.next != null && prev.next.val == val) {
            prev.next = prev.next.next;
        }

        return bak.next;
    }
}

File: Top200/offer18-DeleteNode.go
package main

/**
 * Definition for singly-linked list.
 * type ListNode struct {
 *     Val int
 *     Next *ListNode
 * }
 */
type ListNode struct {
	Val  int
	Next *ListNode
}

func deleteNode(head *ListNode, val int) *ListNode {
	dummy := ListNode{
		Val:  -1,
		Next: head,
	}

	p := &dummy
	for p.Next != nil {
		if p.Next.Val == val {
			p.Next = p.Next.Next
			break
		}
		p = p.Next
	}

	return dummy.Next
}
